src/fixed_rational.jl

```diff
@@ -47,14 +55,28 @@ Base.convert(::Type{F}, x::Rational) where {F<:FixedRational} = F(x)
 Base.convert(::Type{Rational{R}}, x::F) where {R,F<:FixedRational} = Rational{R}(x.num, denom(F))
 Base.convert(::Type{Rational}, x::F) where {F<:FixedRational} = Rational{eltype(F)}(x.num, denom(F))
 Base.convert(::Type{AF}, x::F) where {AF<:AbstractFloat,F<:FixedRational} = convert(AF, x.num) / convert(AF, denom(F))
-Base.convert(::Type{I}, x::F) where {I<:Integer,F<:FixedRational} = convert(I, convert(Rational, x))
+Base.convert(::Type{I}, x::F) where {I<:Integer,F<:FixedRational} = isinteger(x) ? div(x.num, denom(F)) : throw(InexactError(:convert, I, x))
 Base.round(::Type{T}, x::F) where {T,F<:FixedRational} = div(convert(T, x.num), convert(T, denom(F)), RoundNearest)
+Base.decompose(x::F) where {T,F<:FixedRational{T}} = (x.num, zero(T), denom(F))
+
+# Promotion rules:
+Base.promote_rule(::Type{<:FixedRational{T1,den1}}, ::Type{<:FixedRational{T2,den2}}) where {T1,T2,den1,den2} = FixedRational{promote_type(T1,T2),den*den2}
+Base.promote_rule(::Type{<:FixedRational{T1,den}}, ::Type{<:FixedRational{T2,den}}) where {T1,T2,den} = FixedRational{promote_type(T1,T2),den}
+Base.promote_rule(::Type{<:FixedRational{T1}}, ::Type{Rational{T2}}) where {T1,T2} = Rational{promote_type(T1,T2)}
+Base.promote_rule(::Type{<:FixedRational{T1}}, ::Type{T2}) where {T1,T2} = promote_type(Rational{T1}, T2)
+
+# Want to consume integers:
 Base.promote(x::Integer, y::F) where {F<:FixedRational} = (F(x), y)
 Base.promote(x::F, y::Integer) where {F<:FixedRational} = reverse(promote(y, x))
 Base.promote(x::F1, y::F2) where {F1<:FixedRational,F2<:FixedRational} = (@assert denom(F1) == denom(F2); (x, y))
-Base.promote(x, y::F) where {F<:FixedRational} = promote(x, convert(Rational, y))
-Base.promote(x::F, y) where {F<:FixedRational} = reverse(promote(y, x))
-Base.show(io::IO, x::F) where {F<:FixedRational} = show(io, convert(Rational, x))
+
+Base.string(x::FixedRational) =
+    let
+        isinteger(x) && return string(convert(eltype(x), x))
+        g = gcd(x.num, denom(x))
+        return string(div(x.num, g)) * "//" * string(div(denom(x), g))
```

Comment:
Good point. But note that this printing is mostly for the exponential form, like
```julia
julia> u"m^0.5*s^2"
1.0 m¹ᐟ² s²
```
Also, if you copy and paste it exactly, it would again become a FixedRational due to the eager conversion when creating dimension objects:

```julia
julia> u"m^(1//2)*s^2"
1.0 m¹ᐟ² s²

julia> typeof(u"m^(1//2)*s^2")
Quantity{Float64, Dimensions{DynamicQuantities.FixedRational{Int32, 25200}}}
```

Maybe for printing the number by itself we could annotate the numbers similar to FixedPointNumbers.jl. However I should also note that FixedRational is really only set up for this specific use-case, as it lacks many methods required for a general numeric type – so it will convert itself to a Rational if a user tries to do any manual calculations with the dimension values.

(Also, I realize I should have put things like this into a different PR, my mistake... the commits are too intertwined to do a cherry pick unfortunately, I did try)

Reply:
I guess to summarize what I'm saying, I think we actually **want** the user to copy-paste that explicitly and end up with a `Rational` rather than a `FixedRational`.

We wouldn't want the user to work with `FixedRational` manually because it's not set up for general calculations.

Reply:
Sounds good!